packages/ra-core/src/controller/list/ListContext.tsx

```diff
@@ -52,33 +52,6 @@ import { ListControllerResult } from './useListController';
  *     );
  * };
  */
-export const ListContext = createContext<ListControllerResult>({
-    sort: null,
-    data: null,
-    defaultTitle: null,
-    displayedFilters: null,
-    exporter: null,
-    filterValues: null,
-    hasNextPage: null,
-    hasPreviousPage: null,
-    hideFilter: null,
-    isFetching: null,
-    isLoading: null,
-    isPending: null,
-    onSelect: null,
-    onToggleItem: null,
-    onUnselectItems: null,
-    page: null,
-    perPage: null,
-    refetch: null,
-    resource: null,
-    selectedIds: undefined,
-    setFilters: null,
-    setPage: null,
-    setPerPage: null,
-    setSort: null,
-    showFilter: null,
-    total: null,
-});
+export const ListContext = createContext<ListControllerResult | null>(null);
```

Comment:
Having a default value would force choosing between the pending/error/success case, which doesn't make sense e.g. when using `<Datagrid>` outside of  a `ListContext` but with a `data` prop. 

On the other hand, using a `null` default value should mean that `useListContext` returns `ListControllerResult | null`, but for some reason the typeScript inference sees the return value as `ListControllerResult `, even without return type annotation. 

This is convenient, because the alternative is to turn the return type of `useListContext` as `Partial< ListControllerResult>`, which loses in the process the union between loading/error/success types. Or we could have two hooks:

- one that throws if there is no context, but always returns a `ListControllerResult`
- the other that doesn't throw but returns a `Partial< ListControllerResult>`, forcing a lot more runtime checks. 

Having two hooks seems too complex, so let's keep the code as is for now.

Reply:
There aren't many components that actually need the prop fallback: Datagrid, SingleFieldList, and SimpleList. I ended up making a second hook especially for these ones. This allows me to throw an error whenever a `useXXXController` hook is invoked outside of a `XXXContext`, and ensure that the types are correct when reading the context with these hooks.